src/communication/RpcClient.cpp

```diff
@@ -10,3 +10,371 @@
+			                   return true;
+		                   }
+		                   return false;
+	                   }),
+	    c.end());
+
+	// TODO - is there a better way to shrink the internal container?
+	// Maybe instead we should enforce a capacity limit
+	constexpr size_t capacity_shrink_threshold = 16;
+	if ((c.capacity() > capacity_shrink_threshold) &&
+	    (c.size() < c.capacity() / 2)) {
+		c.shrink_to_fit();
+	}
+
+	return all_expired;
+}
+
+// Exposing non-const version so the listen handle can be moved out
+PendingRequest& ScrubablePendingQueue::top() { return c.front(); }
+
+ExpireWorker::ExpireWorker() {
+	worker_ = std::thread([this]() { doWork(); });
+}
+
+ExpireWorker::~ExpireWorker() {
+	stop_ = true;
+	{
+		std::lock_guard lock(pending_mtx_);
+		wake_worker_.notify_one();
+	}
+	worker_.join();
+}
+
+void ExpireWorker::enqueue(PendingRequest&& pending) {
+	std::lock_guard lock(pending_mtx_);
+	pending_.emplace(std::move(pending));
+	wake_worker_.notify_one();
+}
+
+void ExpireWorker::scrub(size_t instance_id) {
+	std::vector<std::function<void(communication::RpcClient::Status)>>
+	    all_expired;
+	{
+		std::lock_guard lock(pending_mtx_);
+		all_expired = pending_.scrub(instance_id);
+		wake_worker_.notify_one();
+	}
+
+	static const v1::UStatus cancel_reason = []() {
+		v1::UStatus reason;
+		reason.set_code(v1::UCode::CANCELLED);
+		reason.set_message("RpcClient for this request was discarded");
+		return reason;
+	}();
+
+	for (auto& expire : all_expired) {
+		expire(cancel_reason);
+	}
+}
+
+void ExpireWorker::doWork() {
+	while (!stop_) {
+		const auto now = std::chrono::steady_clock::now();
+		std::optional<decltype(PendingRequest::expire)> maybe_expire;
+
+		{
+			transport::UTransport::ListenHandle expired_handle;
+			std::lock_guard lock(pending_mtx_);
+			if (!pending_.empty()) {
+				const auto when_expire = pending_.top().when_expire;
+				if (when_expire <= now) {
+					maybe_expire = pending_.top().expire;
+					expired_handle =
+					    std::move(pending_.top().response_listener);
+					pending_.pop();
+				}
+			}
+		}
+
+		if (maybe_expire) {
+			auto& expire = *maybe_expire;
+
+			static const v1::UStatus expire_reason = []() {
+				v1::UStatus reason;
+				reason.set_code(v1::UCode::DEADLINE_EXCEEDED);
+				reason.set_message("Request expired before response received");
+				return reason;
+			}();
+
+			expire(expire_reason);
+
+		} else {
+			std::unique_lock lock(pending_mtx_);
+			if (pending_.empty()) {
+				wake_worker_.wait(
+				    lock, [this]() { return stop_ || !pending_.empty(); });
+			} else {
+				auto wake_when = pending_.top().when_expire;
+				wake_worker_.wait_until(lock, wake_when, [this, &wake_when]() {
+					auto when_next_wake = wake_when;
+					if (!pending_.empty()) {
+						when_next_wake =
+						    std::min(wake_when, pending_.top().when_expire);
+					}
+					return stop_ ||
+					       (std::chrono::steady_clock::now() >= when_next_wake);
+				});
```

Comment:
Comment to explain this better

Reply:
Does this even make sense? It doesn't wake if the queue is empty but it could. And why is it taking the min of wake_when or pending_.top()?

Reply:
Added a comment. It makes more sense now that it has been re-written, too.